frontend/src/selectors/staticContentSelectors.js

```diff
@@ -192,3 +192,43 @@ export const getVarianceStatusOptionsHash = createSelector(
   [getDropdownVarianceStatusOptions],
   createLabelHash
 );
+
+const itemInList = (child, itemToAdd) => {
+  let item = child.find((x) => x.value === itemToAdd.code);
+
+  if (item === undefined) {
+    item = { value: itemToAdd.code, label: itemToAdd.label, children: [] };
+    child.push(item);
+  }
+
+  return item;
+};
+
+const transformMineStatus = (data) => {
+  const statusOptions = [];
+
+  data.forEach((status) => {
+    const code = itemInList(statusOptions, {
+      code: status.mine_operation_status.mine_operation_status_code,
+      label: status.mine_operation_status.description,
+    });
+    if (status.mine_operation_status_reason.mine_operation_status_reason_code !== null) {
+      const reason = itemInList(code.children, {
+        code: status.mine_operation_status_reason.mine_operation_status_reason_code,
+        label: status.mine_operation_status_reason.description,
+      });
+      if (status.mine_operation_status_sub_reason.mine_operation_status_sub_reason_code !== null) {
+        itemInList(reason.children, {
+          code: status.mine_operation_status_sub_reason.mine_operation_status_sub_reason_code,
+          label: status.mine_operation_status_sub_reason.description,
+        });
+      }
+    }
+  });
+  return statusOptions;
```

Comment:
So, I think I understand what's happening here...

We are adding an object (called `code`) to `statusOptions` and then conditionally adding a `reason` property to that object, which has a `children` array property that may contain an object.

If that's right, let me provide some suggestions...

We always add `code` to `statusOptions`, unless an object with the same `code` property already exists in the array. So, I'd suggest initializing `statusOptions = data`. If there are duplicates, you can add a uniqueness filter based on the `code` property. This removes lines 211-214.

Next, since we're not adding/removing elements from the `statusOptions` array anymore, you can just return `statusOptions.map(({ children }) => { // add children + reason here })`

I imagine the whole thing will look something like...
```javascript
const transformMineStatus = (statusOptions) =>
  statusOptions.map((status) => {
    const code = status.mine_operation_status.mine_operation_status_code;
    const label = status.mine_operation_status.description;
    const reason = status.mine_operation_status_sub_reason;
    const subReason = status.mine_operation_status_sub_reason.mine_operation_status_sub_reason_code;

    const reasonChildren = subReason === null ? {} : {
      children: [
        {
          code: subReason.mine_operation_status_sub_reason_code,
          label: subReason.description,
        },
      ],
    };

    const codeChildren = reason === null ? {} : {
      children: [
        {
          code: reason.mine_operation_status_sub_reason_code,
          label: reason.description,
          ...reasonChildren,
        },
      ],
    };

    return {
      code,
      label,
      ...codeChildren,
    };
  });
```

This much more directly shows the output date shape and removes the need for mutations. `itemInList` isn't needed in this version.

If the separate `codeChildren` and `reasonChildren` variables make it confusing, you could just return those nested objects directly instead of pulling them out into variables. That would make the returned data shape maximally explicit. But I think breaking them out makes it a bit nicer to work with.

Reply:
Another option could be to group and map and group and map:
```javascript
let result = records.groupBy((status) => status.mine_operation_status.mine_operation_status_code)
.map((statusCodeGroup) => ({
	value: statusCodeGroup[0].mine_operation_status.mine_operation_status_code,
	label: statusCodeGroup[0].mine_operation_status.description,
	children: statusCodeGroup.groupBy((status) => status.mine_operation_status_reason.mine_operation_status_reason_code)
		.map((statusReasonGroup) => ({
			value:statusReasonGroup[0].mine_operation_status_reason.mine_operation_status_reason_code,
			label: statusReasonGroup[0].mine_operation_status_reason.description,
			children: statusReasonGroup.groupBy((status) => status.mine_operation_status_sub_reason_code)
				.map((statusSubReasonGroup) => ({
					value:statusSubReasonGroup[0].mine_operation_status_sub_reason_code, 
					label: statusSubReasonGroup[0].description
					}))
  }))
}));
```

Reply:
Ryan worked to modify this function into something much more obvious and more focused on using groupby, filter, and map.

Reply:
The new implementation is much easier to follow. Thank you, Justin and Ryan, for making that change.